reana_cluster/backends/kubernetes/k8s.py

```diff
@@ -0,0 +1,510 @@
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+"""Abstract Base Class representing REANA cluster backend."""
+
+import pkg_resources
+from kubernetes.client.rest import ApiException
+
+import json
+import logging
+import yaml
+
+from pkg_resources import parse_version
+
+from kubernetes import client as k8s_client, config as k8s_config
+from kubernetes.client import configuration
+
+from jinja2 import Environment, FileSystemLoader, TemplateNotFound
+
+from reana_cluster import ReanaBackendABC
+
+
+class KubernetesBackend(ReanaBackendABC):
+    # TODO: Docstring
+
+    __cluster_type = 'kubernetes'
+
+    _conf = {
+        'templates_folder': pkg_resources.resource_filename(
+            __name__, '/templates'),
+        'min_version': 'v1.6.4',
+        'max_version': 'v1.6.4',
+    }
+
+    def __init__(self,
+                 cluster_spec,
+                 cluster_conf=None,
+                 kubeconfig=None,
+                 context=None):
+
+        # Load Kubernetes cluster configuration. Defaults to '~/.kube/config'
+        k8s_config.load_kube_config(kubeconfig, context)
+
+        # Instantiate clients for various Kubernetes REST APIs
+        self.corev1api = k8s_client.CoreV1Api()
+        self.versionapi = k8s_client.VersionApi()
+        self.extbetav1api = k8s_client.ExtensionsV1beta1Api()
+
+        self.cluster_spec = cluster_spec
+        self.cluster_conf = self.generate_configuration(cluster_spec)
+
+    @property
+    def cluster_type(self):
+        return self.__cluster_type
+
+    @property
+    def cluster_type(self):
+        return self.cluster_conf
+
+    @property
+    def current_config(self):
+        # TODO: Check what conf holds when `load_kube_config()` hasn't been run
+        if configuration is None:
+            # TODO: Raise somekind of "NotInitialized" error is
+            # load_kube_config() has not been run.
+            pass
+        # TODO: Probably better to return only relevant data not the whole conf
+        return configuration
+
+    @classmethod
+    def generate_configuration(cls, cluster_spec):
+        """Generate Kubernetes backend configuration files to init REANA cluster.
+
+        :param cluster_spec: Dictionary representing complete REANA
+            cluster spec file.
+        """
+
+        # Setup an Environment for Jinja
+        env = Environment(
+            loader=FileSystemLoader(
+                cls._conf['templates_folder']),
+            keep_trailing_newline=False
+        )
+
+        # Define where are backend conf params needed when rendering templates.
+        be_conf_params_fp = cls._conf['templates_folder'] + '/config.yaml'
+
+        try:
+            with open(be_conf_params_fp) as f:
+
+                # Load backend conf params
+                backend_conf_parameters = yaml.load(f.read())
```

Comment:
are we making reference here to the parameters that used to be [here](https://github.com/reanahub/reana-resources-k8s/blob/master/reana_resources_k8s/templates/config.yaml)? If so, we are calling them backend configuration parameters, again it boils down to the naming issue.

If `backend` represents the infrastructure where the `reana-cluster` will be deployed, I feel that it is a bit confusing, shall we maybe name with `cluster` rather than `backend`? what do you think?

Reply:
Could be `cluster_conf_params_fp`.

On the other hand if we change cluster specification file to `cluster`-key to `backend` (as according to reana_cluster_specification = { backend_spec, component_spec(s) }, then this is kind of valid naming.

I think we should chat about cluster specification file content, structure and naming offline and based on the outcome of that discussion we resolve this cluster vs backend naming.